packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/classes.ts

```diff
@@ -206,10 +220,17 @@ export abstract class HandlerFrameworkClass extends ClassType {
           ['handler', expr.lit(props.handler)],
           ['runtime', this.buildRuntimeProperty(scope, { runtime: props.runtime, isEvalNodejsProvider })],
         ]);
+        const metadataStatements: Statement[] = [
+          expr.directCode(`this.addMetadata('${CUSTOM_RESOURCE_SINGLETON}', true)`),
+          expr.directCode(`if (props?.logGroup) { this.logGroup.node.addMetadata('${CUSTOM_RESOURCE_SINGLETON_LOG_GROUP}', true) }`),
+          // (this singleton function) as it extends FunctionBase and property lambdaFunction is available to be cast to lambda.Function to access the internal readonly _logRetention
+          expr.directCode(`if (props?.logRetention) { ((this as any).lambdaFunction as lambda.Function)._logRetention?.node.addMetadata('${CUSTOM_RESOURCE_SINGLETON_LOG_RETENTION}', true) }`),
```

Comment:
I realized we're missing a case here. Look at this code: https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-lambda/lib/function.ts#L1331 and https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-lambda/lib/singleton-lambda.ts#L104.

What we're missing is that a user can go

`const func = new BucketDeploymentSingletonFunction(...)`

`func.logGroup // <-- this call creates a LogRetention custom resource!`

can you check for the `_logGroup` property being set on the lambda, and if it is, add the metadata like you do in the `if (props?.logRetention)` case?

Reply:
Not able to test BucketDeploymentSingletonFunction as it is not exposed to be tested.